### PR TITLE
Access log query enhancements

### LINKED ISSUE #811: Enhancements for Access Log Report
Requirements:
  - [x] Add ability to specify singular `subject` query param that only returns records where `context.subject.label` matches that input
  - [x] Add ability to specify multiple `access_type` query params that only return records where `access_type` matches one of those inputs
  - [ ] Add `csv` query param that, when `True`, returns a csv file of all matching records flattened into rows
  - [x] Write integration tests to thoroughly test the new functionality
  - [x] Create new endpoint (`/api/reports/access/types` ?) that returns the available access types for frontend use.

### PR DESCRIPTION
**Pull request by hkethi002**
Fixes #811 
### Review Checklist

- Tests were added to cover all code changes
- Documentation was added / updated
- Code and tests follow standards in CONTRIBUTING.mdRequirements:

Requirements:
  - [x] Add ability to specify singular `subject` query param that only returns records where `context.subject.label` matches that input
  - [x] Add ability to specify multiple `access_type` query params that only return records where `access_type` matches one of those inputs
  - [x] Add `csv` query param that, when `True`, returns a csv file of all matching records flattened into rows
  - [x] Write integration tests to thoroughly test the new functionality
  - [x] Create new endpoint (`/api/report/accesslog/types` ?) that returns the available access types for frontend use.

### CODE REVIEW

**Comment by nagem** [comment]
Nice work, Harsha. I worry that this endpoint would tip over an instance if asked to create a CSV or return everything in a very large access_log. We'll have to discuss with @ryansanford what limitations for safety seem reasonable and the possibility of a command line tool to generate the CSV when you really do want to have millions of logs returned.

**Comment by ryansanford** [comment]
Somewhere between 70-80k results, we end up with the below error.  That was indeed resolved by adding the index `db.access_log.createIndex({timestamp: -1})`

```
uwsgi_1           | 2017-07-13 02:01:26      scitran.api                  base.py   333:ERRO Traceback (most recent call last):
uwsgi_1           |   File "/usr/local/lib/python2.7/dist-packages/webapp2.py", line 570, in dispatch
uwsgi_1           |     return method(*args, **kwargs)
uwsgi_1           |   File "./api/handlers/reporthandler.py", line 77, in get
uwsgi_1           |     for doc in report.build():
uwsgi_1           |   File "./api/handlers/reporthandler.py", line 566, in build
uwsgi_1           |     return self.make_csv_ready(cursor)
uwsgi_1           |   File "./api/handlers/reporthandler.py", line 545, in make_csv_ready
uwsgi_1           |     return [self.flatten(json_obj, {}) for json_obj in cursor]
uwsgi_1           |   File "/usr/local/lib/python2.7/dist-packages/pymongo/cursor.py", line 1097, in next
uwsgi_1           |     if len(self.__data) or self._refresh():
uwsgi_1           |   File "/usr/local/lib/python2.7/dist-packages/pymongo/cursor.py", line 1019, in _refresh
uwsgi_1           |     self.__read_concern))
uwsgi_1           |   File "/usr/local/lib/python2.7/dist-packages/pymongo/cursor.py", line 905, in __send_message
uwsgi_1           |     helpers._check_command_response(doc['data'][0])
uwsgi_1           |   File "/usr/local/lib/python2.7/dist-packages/pymongo/helpers.py", line 196, in _check_command_response
uwsgi_1           |     raise OperationFailure(msg % errmsg, code, response)
uwsgi_1           | OperationFailure: Executor error during find command: OperationFailed: Sort operation used more than the maximum 33554432 bytes of RAM. Add an index, or specify a smaller limit.
uwsgi_1           |  request_id=c006d926-1499911285
uwsgi_1           | ./api/tempdir.py:42: Warning: start: Implicitly cleaning up <TemporaryDirectory u'/var/scitran/data/.tmpAdZzcl'>
uwsgi_1           |   Warning)
uwsgi_1           | ./api/tempdir.py:57: Warning: end: Implicitly cleaning up <TemporaryDirectory u'/var/scitran/data/.tmpAdZzcl'>
uwsgi_1           |   Warning)
uwsgi_1           | [pid: 46|app: 0|req: 27/121] 172.19.0.1 () {36 vars in 723 bytes} [Thu Jul 13 02:01:25 2017] GET /api/report/accesslog?user=ryansanford@flywheel.io&end_date=2017-07-11T00:00:00.000Z&start_date=2017-06-04T00:00:00.000Z&limit=80000&csv=true => generated 194 bytes in 557 msecs (HTTP/1.1 500) 3 headers in 131 bytes (2 switches on core 0)

```

**Comment by ryansanford** [comment]
Next breakage is happening between 300-400k. uwsgi worker is dying. This is happening after 17 seconds.

```
mongodb_1         | 2017-07-13T02:25:24.840+0000 I NETWORK  [initandlisten] connection accepted from 172.19.0.6:50080 #65 (14 connections now open)
mongodb_1         | 2017-07-13T02:25:24.848+0000 I NETWORK  [initandlisten] connection accepted from 172.19.0.6:50082 #66 (15 connections now open)
uwsgi_1           | 2017-07-13 02:25:24      scitran.api                  base.py   403:DEBU from ryansanford@flywheel.io GET /api/report/accesslog {u'csv': u'true', u'limit': u'400000', u'user': u'ryansanford@flywheel.io', u'end_date': u'2017-07-11T00:00:00.000Z', u'start_date': u'2017-06-04T00:00:00.000Z'} request_id=af740af3-1499912724
uwsgi_1           | 2017-07-13 02:25:24      scitran.api                config.py   223:INFO Initializing database, creating indexes
mongodb_1         | 2017-07-13T02:25:24.885+0000 I NETWORK  [initandlisten] connection accepted from 172.19.0.6:50084 #67 (16 connections now open)
mongodb_1         | 2017-07-13T02:25:24.888+0000 I NETWORK  [initandlisten] connection accepted from 172.19.0.6:50086 #68 (17 connections now open)
uwsgi_1           | 2017-07-13 02:25:24      scitran.api                config.py   255:INFO Persisting configuration
mongodb_1         | 2017-07-13T02:25:39.525+0000 I NETWORK  [conn66] end connection 172.19.0.6:50082 (16 connections now open)
mongodb_1         | 2017-07-13T02:25:39.525+0000 I NETWORK  [conn65] end connection 172.19.0.6:50080 (16 connections now open)
mongodb_1         | 2017-07-13T02:25:39.525+0000 I NETWORK  [conn67] end connection 172.19.0.6:50084 (16 connections now open)
mongodb_1         | 2017-07-13T02:25:39.525+0000 I NETWORK  [conn68] end connection 172.19.0.6:50086 (16 connections now open)
uwsgi_1           | DAMN ! worker 2 (pid: 100) died, killed by signal 9 :( trying respawn ...
uwsgi_1           | Respawned uWSGI worker 2 (new pid: 105)


```

**Comment by ryansanford** [comment]
@hkethi002 Would putting a python script in `bin/` to handle arbitrary large exports make sense and be reliable, or would we have similar issue as uwsgi process dying as seen above?

**Comment by ryansanford** [comment]
Regarding the uwsgi exiting, it's due to running out of memory. It reaches about 2.5 GB at the point of processing 360k results

File: bin/database.py [hunk 1]
```diff
@@ -1164,6 +1164,12 @@ def upgrade_to_34():
     config.db.groups.update_many({'roles': {'$exists': True}}, {'$rename': {'roles': 'permissions'}})
     config.db.groups.update_many({'name': {'$exists': True}}, {'$rename': {'name': 'label'}})
 
+def upgrade_to_35():
+    """
+    Indexes the timestamps of access logs
+    """
+    config.log_db.access_log.createIndex({'timestamp': -1})
```
Comment (nagem) [inline bin/database.py]: You can put this in config.py and it will get added on startup if it's not already there.

**Comment by hkethi002** [comment]
Ran into Unicode Errors caused by some labels having emojis in them, and it seems like csv library being used uses ASCII so should they just be skipped for now?

**Comment by nagem** [comment]
We're going to have to support unicode, especially if we are going to support users whose languages don't use the ASCII char set. Looks like there is a unicode replacement for the csv lib:

https://github.com/jdunck/python-unicodecsv

File: api/handlers/reporthandler.py [hunk 2]
```diff
@@ -460,12 +522,17 @@ def __init__(self, params):
             raise APIReportParamsException('Limit must be an integer greater than 0.')
         if limit < 1:
             raise APIReportParamsException('Limit must be an integer greater than 0.')
+        for access_type in access_types:
+            if access_type not in AccessTypeList:
+                raise APIReportParamsException('Not a valid access type')
```
Comment (nagem) [inline api/handlers/reporthandler.py]: We should decide on a "safe" limit and reject requests that go over that limit. The rejection 400 could mention that an admin should run the script instead. @ryansanford what would you say the safe uwsgi log limit would be after your testing?

Comment (nagem) [inline api/handlers/reporthandler.py]: Then make sure we add a test for failing when attempting to go over that limit.

File: api/handlers/reporthandler.py [hunk 3]
```diff
@@ -460,12 +522,19 @@ def __init__(self, params):
             raise APIReportParamsException('Limit must be an integer greater than 0.')
         if limit < 1:
             raise APIReportParamsException('Limit must be an integer greater than 0.')
+        elif limit > 10000:
+            raise APIReportParamsException('Limit exceeds 1000 entries, please contact admin to run script.')
```
Comment (nagem) [inline api/handlers/reporthandler.py]: Need one more zero on the error :)

**Comment by nagem** [comment]
After error message change, LGTM! Feel free to merge.

**Pull request merged**